src/App.js

```diff
@@ -30,42 +33,98 @@ const App = () => {
   // This starts state off as a 2D array of JS objects with
   // empty value and unique ids.
   const [squares, setSquares] = useState(generateSquares());
+  const [currentPlayer, setCurrentPlayer] = useState(PLAYER_1);
+  const [winner, setWinner] = useState('');
+  const [turns, setTurns] = useState(0)
 
   // Wave 2
   // You will need to create a method to change the square 
   //   When it is clicked on.
   //   Then pass it into the squares as a callback
+  const onClickCallback = (id) => {
+    let row = 0
+    while (id > 2) {
+      id -= 3
+      row++
+    }
 
+    if (!squares[row][id]['value'] && !winner) {
+      const newSquares = [...squares]
```

Comment:
Be mindful that spreading like this only does a shallow copy of the subarrays that were in `squares`. The outer array is new and will trigger a re-render, but the inner arrays are shared with the existing `squares`. This is why the squares reference was able to be used in the `checkForWinner` function and it reflected the updated condition, even though the next render would not yet have occurred. This is a little sneaky, and in general we'd like to avoid doing this (here, by making a deep copy and then passing the modified copy into `checkForWinner` as well) as it may have unanticipated side effects in larger applications.